src/types/resolveStatements.ts

```diff
@@ -353,17 +353,49 @@ function processStatements(
                     // Check type
                     const expressionType = getExpType(ctx, s.expression);
                     const tailType = getExpType(ctx, s.path);
-                    // Check if types are Int
+
+                    // Check if any of the types is not ref or is optional or types themselves don't match
                     if (
                         expressionType.kind !== "ref" ||
-                        expressionType.name !== "Int" ||
-                        expressionType.optional ||
-                        tailType.kind !== "ref" ||
-                        tailType.name !== "Int" ||
-                        tailType.optional
+                        expressionType.optional
                     ) {
                         throwCompilationError(
-                            `Type error: Augmented assignment is only allowed for Int type`,
+                            `Type error: invalid type ${printTypeRef(expressionType)} for augmented assignment`,
+                            s.loc,
+                        );
+                    }
+                    if (tailType.kind !== "ref" || tailType.optional) {
+                        throwCompilationError(
+                            `Type error: invalid type ${printTypeRef(tailType)} for augmented assignment`,
+                            s.loc,
+                        );
+                    }
+
+                    if (s.op === "&&" || s.op === "||") {
+                        if (
+                            expressionType.name !== "Bool" ||
+                            tailType.name !== "Bool"
+                        ) {
+                            throwCompilationError(
+                                `Type error: Augmented assignment ${s.op}= is only allowed for Bool type`,
+                                s.loc,
+                            );
+                        }
+                    } else {
+                        if (
+                            expressionType.name !== "Int" ||
+                            tailType.name !== "Int"
+                        ) {
+                            throwCompilationError(
+                                `Type error: Augmented assignment ${s.op}= is only allowed for Int type`,
+                                s.loc,
+                            );
+                        }
+                    }
+
+                    if (expressionType.name !== tailType.name) {
+                        throwCompilationError(
+                            `Type error: Augmented assignment ${s.op}= is only allowed for matching types`,
```

Comment:
So, at this point we now the l-value and the operator are all good, so we need to complain about the type of the expression. We should be a bit more precise both in the error message and the location. Also, let's add a negative test for this error message.

For instance,

```
let x: Int = 42;
x += true;
```

`x: Int` and `+=` are matching, so we just report that `true` is expected to have type `Int` with the location pointer pointing at `true`, not at the whole statement.

Reply:
actually it seems that we'll never throw this error, because both types are being checked above already. have a look at the most recent commit.